Lib/ufo2ft/__init__.py

```diff
@@ -872,6 +928,62 @@ def _compileNeededSources(
 
         # Stick TTFs back into original big DS
         for ttfSource in ttfDesignSpace.sources:
+            if can_optimize_features:
+                originalSources[ttfSource.name] = sourcesByName[ttfSource.name].font
             sourcesByName[ttfSource.name].font = ttfSource.font
 
-    return vfNameToBaseUfo
+    return vfNameToBaseUfo, can_optimize_features, originalSources
+
+
+def compile_all_variable_features(
+    designSpaceDoc, vfNameToTTFont, originalSources, debugFeatureFile=False
+):
+    interpolableSubDocs = [
+        subDoc for _location, subDoc in splitInterpolable(designSpaceDoc)
+    ]
+    for subDoc in interpolableSubDocs:
+        for vfName, vfDoc in splitVariableFonts(subDoc):
+            if vfName not in vfNameToTTFont:
+                continue
+            ttFont = vfNameToTTFont[vfName]
+            # vfDoc is now full of TTFs, create a UFO-sourced equivalent
+            ufoDoc = vfDoc.deepcopyExceptFonts()
+            for ttfSource, ufoSource in zip(vfDoc.sources, ufoDoc.sources):
+                ufoSource.font = originalSources[ttfSource.name]
+            compile_variable_features(ufoDoc, ttFont, debugFeatureFile)
+
+
+def compile_variable_features(designSpaceDoc, ttFont, debugFeatureFile):
+    default_ufo = designSpaceDoc.findDefault().font
+
+    # Delete anything from the UFO glyphset which didn't make it into the font.
+    fontglyphs = ttFont.getGlyphOrder()
+    glyphSet = {g.name: g for g in default_ufo if g.name in fontglyphs}
+
+    # Add anything we added to the TTF without telling the UFO
+    if ".notdef" not in glyphSet:
+        glyphSet[".notdef"] = StubGlyph(".notdef", 0, 0, 0, 0)
```

Comment:
I don't like that we have to do this here.. Ideally we would pass down the same glyphSet that was preprocessed and potentially expanded by the outlineCompiler

Reply:
Ideally, yes. But that would require having the glyphsets handy in `compileVariableTTFs`, which would mean returning them from `_compileNeededSources`, which would mean returning them from `compileInterpolatableTTFsFromDS`, which would mean returning them from `compileInterpolatableTTFs`. Major surgery.

Reply:
yes, it can't be otherwise

Reply:
the default master's glyphSet is by definition a superset of the glyphsets of the non default sources